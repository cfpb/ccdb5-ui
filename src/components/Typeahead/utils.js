/**
 * Sets the options for the async typeahead
 *
 * @param {string} value - value typed into the typeahead
 * @param {Function} setOptions - function to set local array of options
 * @param {string} uri - the endpoint being hit by the fetch
 * @returns {void | Promise<object>} - Either nothing or a response with option object
 */
export function handleFetchSearch(value, setOptions, uri) {
  const num = value.toLowerCase();
  if (num === '') {
    setOptions([]);
    return;
  }
  return fetch(uri)
    .then((result) => result.json())
    .then((items) => {
      const options = items.map((item) => ({
        key: item,
        label: item,
        position: item.toLowerCase().indexOf(num),
        value,
      }));
      setOptions(options);
    });
}
