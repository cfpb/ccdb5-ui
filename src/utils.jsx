import { DATE_RANGE_MIN, SLUG_SEPARATOR } from './constants'
import moment from 'moment'

// eslint-disable-next-line complexity
export const calculateDateInterval = ( minDate, maxDate ) => {
  // only check intervals if the end date is today
  // round off the date so the partial times don't mess up calculations
  const today = moment( new Date() ).startOf( 'day' )
  const end = moment( maxDate ).startOf( 'day' )
  const start = moment( minDate ).startOf( 'day' )

  // make sure end date is the same as today's date
  if ( end.diff( today, 'days', true ) !== 0 ) {
    return ''
  }

  // is the start date the same as the oldest document?
  if ( moment( minDate ).isSame( DATE_RANGE_MIN, 'day' ) ) {
    return 'All'
  }

  // verify if it's 3 or 1 years
  const yrDiff = end.diff( start, 'years', true )

  if ( yrDiff === 3 || yrDiff === 1 ) {
    return yrDiff + 'y'
  }

  // verify if it's 6 or 3 months
  const moDiff = end.diff( start, 'months', true )
  if ( moDiff === 6 || moDiff === 3 ) {
    return moDiff + 'm'
  }

  return ''
}

/**
 * Function to set the limit of the range of a set of numbers
 * @param {int} x value we are checking
 * @param {int} min smallest number it can me
 * @param {int} max biggest number it can be
 * @returns {*}the limited value
 */
export const clamp = ( x, min, max ) => {
  if ( x < min ) {
    x = min;
  } else if ( x > max ) {
    x = max;
  }
  return x;
}

/**
 * Replacement for the common pattern:
 * if( o.field )
 *    x = o.field
 * else
 *    x = alternateValue
 *
 * Avoids some of the complexity lint warnings
 *
 * @param {Object} o the object being tested
 * @param {string} field the field to check
 * @param {string|Object} alternateValue the value to use in absence
 * @returns {string} the value to use
 */
export const coalesce = ( o, field, alternateValue ) => {
  if ( typeof o !== 'object' ) {
    return alternateValue;
  }

  return field in o && o[field] ? o[field] : alternateValue;
};

/**
 * Creates a hash from a string
 *
 * @param {string} someString the string to hash
 * @returns {int} a hashing of the string
 */
export function hashCode( someString ) {
  const s = String( someString )
  let hash = 0
  let i, chr
  if ( s.length === 0 ) {
    return hash
  }
  for ( i = 0; i < s.length; i++ ) {
    chr = s.charCodeAt( i )
    hash = ( hash << 5 ) - hash + chr

    // Convert to 32bit integer
    hash |= 0
  }
  return hash
}

/**
 * Creates a hash from an object
 *
 * @param {string} o the object to hash
 * @returns {string} a hashing of the object
 */
export function hashObject( o ) {
  return hashCode( JSON.stringify( o ) )
}
export const normalize = s => s.toLowerCase()

export const slugify = ( a, b ) => a + SLUG_SEPARATOR + b

export const sortSelThenCount = ( options, selected ) => {
  const retVal = ( options || [] ).slice()

  // Sort the array so that selected items appear first, then by doc_count
  /* eslint complexity: ["error", 5] */
  retVal.sort( ( a, b ) => {
    const aSel = selected.indexOf( a.key ) !== -1
    const bSel = selected.indexOf( b.key ) !== -1

    if ( aSel && !bSel ) {
      return -1;
    }
    if ( !aSel && bSel ) {
      return 1;
    }

    // Both are selected or not selected
    // Sort by descending doc_count
    return b.doc_count - a.doc_count
  } )

  return retVal
}

/**
* Safely format a date
*
* @param {Date} date the date to convert
* @returns {string} the date formatted for the current locale
*/
export function shortFormat( date ) {
  const wrapped = moment( date )
  return date ? wrapped.format( 'M/D/YYYY' ) : ''
}

/**
* Convert a date to a truncated ISO-8601 string
*
* @param {Date} date the date to convert
* @returns {string} the date formatted as yyyy-mm-ddd
*/
export function shortIsoFormat( date ) {
  return date ? date.toISOString().substring( 0, 10 ) : ''
}

// ----------------------------------------------------------------------------
// attribution: underscore.js (MIT License)

/**
* Delay the implementation of a function until after a period of time
* This prevents expensive calls from being made while triggering events are
* still happening
*
* @param {function} func a function with an embedded expensive call
* @param {int} wait the number of msecs to delay before calling the function
* @returns {function} a replacement function to use in place of the original
*/
export function debounce( func, wait ) {
  var timer = null;

  var later = ( context, args ) => {
    timer = null;
    func.apply( context, args );
  }

  return ( ...theArgs ) => {
    if ( !timer ) {
      timer = setTimeout( later, wait, this, theArgs )
    }
  }
}

// ----------------------------------------------------------------------------
// attribution: lodash.js (Creative Commons License)

/**
* Binds methods of an object to the object itself, overwriting the existing
* method
*
* @param {Object} obj The object to bind and assign the bound methods to.
* @param {...(string|string[])} methodNames The object method names to bind,
*  specified individually or in arrays.
* @returns {Object} the updated object
*/
export function bindAll( obj, methodNames ) {
  const length = methodNames.length
  for ( let i = 0; i < length; i++ ) {
    const key = methodNames[i]
    obj[key] = obj[key].bind( obj )
  }
  return obj;
}
