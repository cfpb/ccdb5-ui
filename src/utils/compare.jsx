export const arrayEquals = (a, b) => {
  let i = a.length;
  if (i !== b.length) return false;
  while (i--) {
    if (a[i] !== b[i]) return false;
  }
  return true;
};
