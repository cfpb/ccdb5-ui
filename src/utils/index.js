import {
  DATE_RANGE_MIN,
  flagFilters,
  knownFilters,
  SLUG_SEPARATOR,
} from '../constants/index';
import Analytics from '../actions/analytics';
import dayjs from 'dayjs';
import queryString from 'query-string';

/**
 * Breaks up '123' to '1 2 3' to help screen readers read digits individually
 * https://thatdevgirl.com/blog/accessibility-phone-number-formatting
 *
 * @param {string} digits - the string of digits
 * @returns {string} an expanded string of digits
 */
export function ariaReadoutNumbers(digits) {
  return Array.from(digits || '').join(' ');
}

/* eslint-disable no-console */

// eslint-disable-next-line complexity
export const calculateDateRange = (minDate, maxDate) => {
  // only check intervals if the end date is today
  // round off the date so the partial times don't mess up calculations
  const today = dayjs(startOfToday());
  const end = dayjs(maxDate).startOf('day');
  const start = dayjs(minDate).startOf('day');

  // make sure end date is the same as today's date
  if (end.diff(today, 'days') !== 0) {
    return '';
  }

  // is the start date the same as the oldest document?
  if (dayjs(minDate).isSame(DATE_RANGE_MIN, 'day')) {
    return 'All';
  }

  // verify if it's 3 or 1 years
  const yrDiff = end.diff(start, 'years', true);
  if (yrDiff === 3 || yrDiff === 1) {
    return yrDiff + 'y';
  }

  // verify if it's 6 or 3 months
  const moDiff = end.diff(start, 'months', true);
  if (moDiff === 6 || moDiff === 3) {
    return moDiff + 'm';
  }

  return '';
};

/**
 * Clamps number within the inclusive lower and upper bounds.
 * https://github.com/you-dont-need/You-Dont-Need-Lodash-Underscore#_clamp
 *
 * @param {number} number - The value we are checking.
 * @param {number} boundOne - The lower bound we don't want to go before.
 * @param {number} boundTwo - The upper bound we can't go past.
 * @returns {number} The clamped number.
 */
export const clamp = (number, boundOne, boundTwo) => {
  if (!boundTwo) {
    return Math.max(number, boundOne) === boundOne ? boundOne : number;
  } else if (Math.min(number, boundOne) === number) {
    return boundOne;
  } else if (Math.max(number, boundTwo) === number) {
    return boundTwo;
  }
  return number;
};

/**
 * Function to set the limit of the range of a set of dates
 *
 * @param {string} val - value we are checking
 * @param {string} min - smallest number it can be
 * @param {string} max - biggest number it can be
 * @returns {*} the limited value
 */
export const clampDate = (val, min, max) => {
  let xDate = new Date(val);
  const minDate = new Date(min);
  const maxDate = new Date(max);

  if (xDate < minDate) {
    xDate = minDate;
  } else if (xDate > maxDate) {
    xDate = maxDate;
  }
  return xDate;
};

/**
 * Replacement for the common pattern:
 * if( o.field )
 * x = o.field
 * else
 * x = alternateValue
 *
 * Avoids some of the complexity lint warnings
 *
 * @param {object} object - the object being tested
 * @param {string} field - the field to check
 * @param {string | object} alternateValue - the value to use in absence
 * @returns {string | Array | object} the value to use
 */
export const coalesce = (object, field, alternateValue) => {
  if (typeof object !== 'object') {
    return alternateValue;
  }

  return field in object && object[field] ? object[field] : alternateValue;
};

/**
 * Creates a hash from a string
 *
 * @param {string} someString - the string to hash
 * @returns {number} a hashing of the string
 */
export function hashCode(someString) {
  const str = String(someString);
  let hash = 0;
  let index, chr;
  if (str.length === 0) {
    return hash;
  }
  for (index = 0; index < str.length; index++) {
    chr = str.charCodeAt(index);
    hash = (hash << 5) - hash + chr;

    // Convert to 32bit integer
    hash |= 0;
  }
  return hash;
}

/**
 * helper function to determine if we have any filters selected so we can
 * disable the Per 1000 Complaints button
 * enable per1000 if the only filter selected is state
 *
 * @param {object} filters - reducer contains values for the filters, etc
 * @returns {boolean} are we enabling the perCap
 */
// eslint-disable-next-line complexity
export function enablePer1000(filters) {
  const keys = [];
  let filter;
  const allFilters = knownFilters.concat(flagFilters);

  for (let index = 0; index < allFilters.length; index++) {
    filter = allFilters[index];

    if (
      (Array.isArray(filters[filter]) && filters[filter].length) ||
      filters[filter] === true
    ) {
      keys.push(filter);
    }
  }
  const compReceivedFilters = ['company_received_max', 'company_received_min'];
  for (let index = 0; index < compReceivedFilters.length; index++) {
    filter = compReceivedFilters[index];
    if (filters[filter]) {
      keys.push(filter);
    }
  }

  if (keys.length) {
    // normalization still okay as long as only State filters are applied
    return keys.length === 1 && keys[0] === 'state';
  }

  return true;
}

export const normalize = (str) => str.toLowerCase();

/**
 * takes a string and formats it into proper text for an htmd ID
 * Eat at Joe's => eatatjoes
 *
 * @param {string} str - the dirty string Eat at Joe's
 * @returns {string} sanitized string eat-at-joe-s
 */
export const sanitizeHtmlId = (str) =>
  str.replace(/\s+|\W/g, '-').toLowerCase();

export const slugify = (first, second) => first + SLUG_SEPARATOR + second;

/**
 * Custom sort for array so that selected items appear first, then by doc_count
 *
 * @param {Array} options - input array containing values
 * @param {Array} selected - values
 * @returns {Array} sorted array
 */
export const sortSelThenCount = (options, selected) => {
  const retVal = (structuredClone(options) || []).slice();

  /* eslint complexity: ["error", 5] */
  retVal.sort((first, second) => {
    const aSel = selected.indexOf(first.key) !== -1;
    const bSel = selected.indexOf(second.key) !== -1;

    if (aSel && !bSel) {
      return -1;
    }
    if (!aSel && bSel) {
      return 1;
    }

    // Both are selected or not selected
    // Sort by descending doc_count
    return second.doc_count - first.doc_count;
  });

  return retVal;
};

/**
 * Safely format a date
 *
 * @param {Date} date - the date to convert
 * @returns {string} the date formatted for the current locale
 */
export function shortFormat(date) {
  const wrapped = dayjs(date);
  return date ? wrapped.format('M/D/YYYY') : '';
}

/**
 * Convert a date to a truncated ISO-8601 string
 *
 * @param {string | object | Date} date - the date to convert
 * @returns {string} the date formatted as yyyy-mm-ddd
 */
export function shortIsoFormat(date) {
  if (typeof date === 'string') {
    return date.slice(0, 10);
  } else if (typeof date === 'object' && date !== null) {
    return dayjs(date).toISOString().slice(0, 10);
  }
  return '';
}

/**
 * Gets the UTC time for the beginning of the day in the local time zone
 *
 * @returns {string} midnight today, local
 */
export function startOfToday() {
  if (!Object.prototype.hasOwnProperty.call(window, 'MAX_DATE')) {
    if (
      Object.prototype.hasOwnProperty.call(window, 'complaint_public_metadata')
    ) {
      const { metadata_timestamp: stamp } = window.complaint_public_metadata;
      window.MAX_DATE = new Date(dayjs(stamp).startOf('day').toString());
    } else {
      console.error('complaint_public_metadata is missing');
      window.MAX_DATE = new Date(dayjs().startOf('day').toString());
    }
  }

  // Always return a clone so the global is not exposed or changed
  return new Date(window.MAX_DATE.valueOf());
}

// ----------------------------------------------------------------------------
// attribution: underscore.js (MIT License)

/**
 * Native implementation of lodash debounce
 * https://github.com/you-dont-need/You-Dont-Need-Lodash-Underscore#_debounce
 *
 * @param {Function} func - The function to run.
 * @param {number} wait - Time in milliseconds.
 * @param {boolean} immediate - Whether we should run function immedately.
 * @returns {Function} the debounced function
 */
export function debounce(func, wait, immediate) {
  let timeout;
  return function () {
    const context = this,
      args = arguments;
    clearTimeout(timeout);
    timeout = setTimeout(function () {
      timeout = null;
      if (!immediate) func.apply(context, args);
    }, wait);
    if (immediate && !timeout) func.apply(context, args);
  };
}

/**
 * Makes sure that a URI has host, protocol, etc.
 *
 * @param {string} uri - the uri to test
 * @returns {string} a uri with the protocol, host and port if necessary
 */
export function getFullUrl(uri) {
  if (!uri) return uri;

  // https://gist.github.com/jlong/2428561
  const parser = document.createElement('a');
  parser.href = uri;
  return parser.href;
}

/**
 * processes error messages so we can see them in redux
 *
 * @param {Error} err - the error object from api
 * @returns {{name: string, message: string}} processed error object we can see
 */
export function processErrorMessage(err) {
  return {
    name: err.name,
    message: err.message,
  };
}

/**
 * Takes in a number and outputs to percentage
 *
 * @param {number} num - value we convert .9999
 * @returns {number} 99.99
 */
export function formatPercentage(num) {
  // we have to do this so it is a float and not a string
  const val = parseFloat(parseFloat(num * 100).toFixed(2));
  return isNaN(val) ? 0.0 : val;
}

/**
 * helper function
 *
 * @param {object} bucket - contains key value pairs
 * @returns {string} name of the key that has the buckets
 */
export const getSubKeyName = (bucket) => {
  for (const item in bucket) {
    if (item !== 'trend_period' && bucket[item].buckets) {
      return item;
    }
  }
  return '';
};

/**
 * helper function to take in array parameters from the url, filters, etc and
 * set the values in the processed object
 *
 * @param {object} params - the object from the URL_CHANGED action
 * @param {object} state - the state we will update with a single value or arr
 * @param {object} arrayParams - the array of strings that we will check against
 */
export const processUrlArrayParams = (params, state, arrayParams) => {
  arrayParams.forEach((field) => {
    if (typeof params[field] !== 'undefined') {
      if (typeof params[field] === 'string') {
        state[field] = [params[field]];
      } else {
        state[field] = params[field];
      }
    }
  });

  if (params.has_narrative) {
    state.has_narrative = !!params.has_narrative;
  } else {
    delete state.has_narrative;
  }
};

/**
 * gets a filter and its subagg filters
 *
 * @param {string} filterKey - the filter 'Debt'
 * @param {Array} subitems - the buckets to process to generate slug
 * @returns {Set<any>} returns a set of uniques Debt, Debt*Foo
 */
export const getAllFilters = (filterKey, subitems) => {
  const values = new Set();
  // Add the parent
  values.add(filterKey);
  // Add the shown subitems
  subitems.forEach((sub) => {
    values.add(slugify(filterKey, sub.key));
  });
  return values;
};

/**
 * Wrapper around analytics event action creator to minimize the copypasta
 *
 * @param {string} action - GA Action (not redux action)
 * @param {string} label - param used by GA
 */
export const sendAnalyticsEvent = (action, label) => {
  Analytics.sendEvent(Analytics.getDataLayerOptions(action, label));
};

/**
 * Helper function to get the selected class based on two different values
 *
 * @param {string|number|boolean} first - Value 1 to compare
 * @param {string|number|boolean} second - Value 2 to compare
 * @param {string} selectedClassName - The value that should be returned if both are the same
 * @returns {string} The selected class
 */
export const selectedClass = (
  first,
  second,
  selectedClassName = 'selected',
) => {
  return first === second ? ' ' + selectedClassName : '';
};

/**
 * Remove all properties with the value 'null' from the object, or empty string
 *
 * @param {object} object - the object with potential nulls
 * @returns {object} the processed object
 */
export function removeNullProperties(object) {
  const myObject = Object.keys(object).reduce((acc, key) => {
    if (
      object[key] !== null &&
      object[key] !== undefined &&
      object[key] !== '' &&
      !Number.isNaN(object[key])
    ) {
      acc[key] = object[key];
    }
    return acc;
  }, {});

  for (const key in myObject) {
    if (Array.isArray(myObject[key]) && myObject[key].length === 0) {
      delete myObject[key];
    }
  }

  return myObject;
}

/**
 * Builds a URL from a path and dictionary
 *
 * @param {string} path - The V2 endpoint.
 * @param {object} params - A key/value pair of the query string params.
 * @returns {string} The full endpoint url.
 */
export function formatUri(path, params) {
  return path + '?' + queryString.stringify(params);
}
