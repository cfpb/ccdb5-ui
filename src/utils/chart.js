// ----------------------------------------------------------------------------
/* eslint-disable complexity */
import { adjustDate, isDateEqual, formatDisplayDate } from './formatDate';
import { clampDate, coalesce } from '../utils';
import dayjs from 'dayjs';
import dayjsQuarterOfYear from 'dayjs/plugin/quarterOfYear';
import dayjsTimezone from 'dayjs/plugin/timezone';
import dayjsUtc from 'dayjs/plugin/utc';
import * as colors from '../constants/colors';

dayjs.extend(dayjsQuarterOfYear);
dayjs.extend(dayjsUtc);
dayjs.extend(dayjsTimezone);
// this is how we enforce standard tooltip.
// ci has Africa/Abidjan
dayjs.tz.setDefault('America/New_York');

export const getLastDate = (dataSet, config) => {
  // take in array of data points
  // early exit
  if (!dataSet || dataSet.length === 0) {
    return null;
  }

  const deDuped = [
    ...new Set(dataSet.map((obj) => dayjs(obj.date).toISOString())),
  ].sort();
  const lastDate = deDuped.pop();
  const lastPointValues = dataSet.filter((obj) =>
    isDateEqual(obj.date, lastDate),
  );
  return {
    key: lastDate,
    date: lastDate,
    dateRange: config.dateRange,
    interval: config.interval,
    values: lastPointValues,
  };
};

export const getLastLineDate = (dataSet, config) => {
  // take in array of data points
  if (!dataSet || !dataSet.dataByTopic || dataSet.dataByTopic.length === 0) {
    return null;
  }

  let dates = [];
  dataSet.dataByTopic.forEach((datum) => {
    dates = dates.concat(datum.dates);
  });

  const deDuped = [...new Set(dates.map((obj) => obj.date))].sort();
  const lastDate = deDuped.pop();
  const values = dataSet.dataByTopic.map((datum) => {
    const lastPoint = datum.dates.find((val) =>
      isDateEqual(val.date, lastDate),
    );
    const value = lastPoint ? lastPoint.value : 0;
    return {
      name: datum.topic,
      date: lastDate,
      value,
    };
  });

  const lastPoint = {
    key: lastDate,
    date: lastDate,
    dateRange: config.dateRange,
    interval: config.interval,
    values,
  };
  return lastPoint;
};

export const getTooltipDate = (inputDate, dateRange) => {
  const adjustedDate = adjustDate(inputDate);
  const returnDate = clampDate(adjustedDate, dateRange.from, dateRange.to);
  return formatDisplayDate(returnDate);
};

export const getTooltipTitle = (inputDate, interval, dateRange, external) => {
  /* eslint complexity: ["error", 6] */
  interval = interval.toLowerCase();
  const startDate = getTooltipDate(inputDate, dateRange);

  let endDate = dayjs(inputDate).utc();

  switch (interval) {
    case 'day':
      endDate = endDate.format();
      break;
    case 'week':
    case 'year':
      endDate = endDate.add(1, interval).subtract(1, 'day').format();
      break;
    case 'quarter':
    case 'month':
    default:
      endDate = endDate.endOf(interval).subtract(1, 'day').format();
      break;
  }

  endDate = getTooltipDate(endDate, dateRange);

  if (interval === 'day') {
    return `Date: ${endDate}`;
  }

  return external
    ? `Date range: ${startDate} - ${endDate}`
    : `${startDate} - ${endDate}`;
};

/**
 * helper to generate chart name for row chart based on trends
 *
 * @param {Array} rowNames - passed from trends.results reducer
 * @param {object} colorMap - of the chart for display
 * @param {string} lens - determines which colors to use for defaults
 * @returns {Array} the color scheme [blue, red, yellow, etc]
 */
export const getColorScheme = (rowNames, colorMap, lens) =>
  rowNames.map((obj) => {
    if (!colorMap) {
      return '#20aa3f';
    }
    // bad data. Some titles can appears twice in the product data
    const name = obj.name.trim();
    const parent = obj.parent ? obj.parent.trim() : '';
    // parent should have priority
    if (colorMap[parent]) {
      return colorMap[parent];
    } else if (colorMap[name]) {
      return colorMap[name];
    }

    // return default grey when it's data lens and not in area/line chart
    // #a2a3a4
    return lens === 'Overview' ? '#20aa3f' : '#a2a3a4';
  });

/**
 * helper function to get d3 bar chart data
 *
 * @param {object} obj - rowdata we are processing
 * @param {Array} nameMap - list of names we are keeping track of
 * @returns {object} the rowdata for row chart
 */
export const getD3Names = (obj, nameMap) => {
  let name = obj.key;
  // D3 doesnt allow dupe keys, so we have to to append
  // spaces so we have unique keys
  while (nameMap[name]) {
    name += ' ';
  }

  nameMap[name] = true;

  return obj.splitterText
    ? obj
    : {
        hasChildren: Boolean(obj.hasChildren),
        isNotFilter: false,
        isParent: Boolean(obj.isParent),
        name: name,
        value: Number(obj.doc_count),
        parent: obj.parent || false,
        // this adjusts the thickness of the parent or child bars
        width: obj.parent ? 0.4 : 0.5,
      };
};

export const processRows = (data, colorMap, lens, expandedRows) => {
  const rows = structuredClone(data);
  if (rows) {
    let data = rows;
    data = data.filter(
      (datum) => datum.isParent || expandedRows.includes(datum.parent),
    );
    const colorScheme = getColorScheme(data, colorMap, lens);

    return {
      colorScheme,
      data,
    };
  }

  return {
    colorScheme: [],
    data: [],
  };
};

/**
 * The api sends us the date buckets in older -> new order
 * however, in data lens / company aggregation it's reversed
 * we also need to fill any empty area buckets when dates are missing
 *
 * @param {string} name - bucket name
 * @param {Array} buckets - contains dates and value paired objects
 * @param {Array} areaBuckets - the reference dates we check against
 * @returns {Array} the sorted array in old-> newest
 */
export const updateDateBuckets = (name, buckets, areaBuckets) => {
  // fill in empty zero values
  areaBuckets.forEach((obj) => {
    if (!buckets.find((bucket) => bucket.key_as_string === obj.key_as_string)) {
      buckets.push({
        name: name,
        doc_count: 0,
        key_as_string: obj.key_as_string,
      });
    }
  });

  return buckets

    .sort((first, second) =>
      first.key_as_string > second.key_as_string ? 1 : -1,
    )
    .map((obj) => ({
      name: name,
      date: obj.key_as_string,
      value: obj.doc_count,
    }));
};

export const externalTooltipFormatter = (tooltip, colorMap) => {
  if (!tooltip) {
    return tooltip;
  }
  const newTooltip = structuredClone(tooltip);
  const parts = tooltip.title.split(':');
  const colorValues = Object.values(colors.DataLens);
  newTooltip.values.forEach((obj) => {
    // const newObj = { ...obj };
    if (!Object.hasOwn(obj, 'colorIndex')) {
      obj.colorIndex = colorValues.indexOf(colorMap[obj.name]) || 0;
    }
    // make sure all values have a value
    if (!Object.hasOwn(obj, 'value')) {
      obj.value = coalesce(obj, 'value', 0);
    }
  });

  return {
    ...newTooltip,
    heading: parts[0] + ':',
    date: parts[1] ? parts[1].trim() : '',
  };
};

export const dateOutOfStartBounds = (dateFrom, startFromChart, interval) => {
  const completeStartPeriod = dayjs(startFromChart)
    .utc()
    .startOf(interval.toLowerCase());
  const dateRangeFrom = dayjs(dateFrom).utc();
  const isSameFrom = dateRangeFrom.isSame(completeStartPeriod, 'day');
  return !isSameFrom;
};

export const dateOutOfEndBounds = (dateTo, lastFromChart, interval) => {
  const completeEndPeriod = dayjs(lastFromChart)
    .utc()
    .endOf(interval.toLowerCase());
  const dateRangeTo = dayjs(dateTo).utc();
  const isSameTo = dateRangeTo.isSame(completeEndPeriod, 'day');
  const afterEnd = completeEndPeriod.isAfter(dateRangeTo);

  return afterEnd && !isSameTo;
};

/**
 * Helper function to determine if the array passed has no dates
 *
 * @param {object} data - Contains an object for the LineCharts.
 * @returns {boolean} Tells us whether we have dates to render the chart.
 */
export const isLineDataEmpty = (data) => {
  return (
    !data ||
    !data.dataByTopic ||
    !data.dataByTopic.length ||
    !data.dataByTopic[0].dates.length ||
    // we consider line data to be empty if length < 2
    // since you need at least 2 points to plot
    data.dataByTopic[0].dates.length < 2
  );
};

/**
 * Helper function to determine if the array passed does not have enough dates
 *
 * @param {Array} data - Contains an array of objects for the StackedAreaCharts.
 * @returns {boolean} Tells us whether we have dates to render the chart.
 */
export const isStackedAreaDataEmpty = (data) => {
  if (!data || !data.length) {
    return true;
  }
  const allDates = [...new Set(data.map((obj) => obj.date))];
  return allDates.length < 2;
};

/**
 * Removes incomplete date periods from the line chart
 *
 * @param {object} data - Object containing arrays to render a line chart
 * @param {object} dateRange - Object containing dateFrom and dateTo
 * @param {string} interval - Month, Day, Year, Quarter, etc
 * @returns {object} Object containing parameters to render the line charts.
 */
export const pruneIncompleteLineInterval = (data, dateRange, interval) => {
  const dataClone = structuredClone(data);
  const { from: dateFrom, to: dateTo } = dateRange;
  if (!dataClone.dataByTopic) {
    return data;
  }

  const dates = dataClone.dataByTopic[0].dates;
  // date from chart
  const startFromChart = dataClone.dataByTopic[0].dates[0].date;
  const lastFromChart = dataClone.dataByTopic[0].dates[dates.length - 1].date;

  // start date from chart same as date range from, then go ahead keep it
  if (dateOutOfStartBounds(dateFrom, startFromChart, interval)) {
    dataClone.dataByTopic.forEach((datum) => {
      datum.dates = datum.dates.filter((date) => date.date !== startFromChart);
    });
  }

  // we only eliminate the last incomplete interval
  // this is if the end date of the interval comes after To Date
  if (dateOutOfEndBounds(dateTo, lastFromChart, interval)) {
    dataClone.dataByTopic.forEach((datum) => {
      datum.dates = datum.dates.filter((date) => date.date !== lastFromChart);
    });
  }
  return dataClone;
};

export const pruneIncompleteStackedAreaInterval = (
  data,
  dateRange,
  interval,
) => {
  const { from: dateFrom, to: dateTo } = dateRange;
  let filteredData = structuredClone(data);
  //  need to rebuild and sort dates in memory
  const dates = [...new Set(filteredData.map((datum) => datum.date))];
  dates.sort();

  const startFromChart = dates[0];
  const lastFromChart = dates[dates.length - 1];
  // start date from chart same as date range from, then go ahead keep it
  if (dateOutOfStartBounds(dateFrom, startFromChart, interval)) {
    filteredData = filteredData.filter(
      (datum) => datum.date !== startFromChart,
    );
  }

  if (dateOutOfEndBounds(dateTo, lastFromChart, interval)) {
    filteredData = filteredData.filter((datum) => datum.date !== lastFromChart);
  }

  return filteredData;
};
