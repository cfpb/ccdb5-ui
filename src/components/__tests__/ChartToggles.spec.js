import {
  mapDispatchToProps,
  mapStateToProps,
  ChartToggles,
} from '../RefineBar/ChartToggles';
import React from 'react';
import { shallow } from 'enzyme';
import * as utils from '../../utils';

/**
 *
 * @param cb
 * @param chartType
 */
function setupEnzyme(cb, chartType) {
  return shallow(<ChartToggles toggleChartType={cb} chartType={chartType} />);
}

describe('component: ChartToggles', () => {
  describe('buttons', () => {
    let cb = null;
    let target = null;

    beforeEach(() => {
      cb = jest.fn();
    });

    it('Line - changeChartType is called the button is clicked', () => {
      target = setupEnzyme(cb, 'foo');
      const prev = target.find('.chart-toggles .line');
      prev.simulate('click');
      expect(cb).toHaveBeenCalledWith('line');
    });

    it('Area - changeChartType is called the button is clicked', () => {
      target = setupEnzyme(cb, 'foo');
      const prev = target.find('.chart-toggles .area');
      prev.simulate('click');
      expect(cb).toHaveBeenCalledWith('area');
    });

    it('changeChartType is NOT called when chartType is the same', () => {
      target = setupEnzyme(cb, 'line');
      const prev = target.find('.chart-toggles .line');
      prev.simulate('click');
      expect(cb).not.toHaveBeenCalled();
    });
  });

  describe('mapDispatchToProps', () => {
    it('provides a way to call changeChartType', () => {
      const dispatch = jest.fn();
      const gaSpy = jest.spyOn(utils, 'sendAnalyticsEvent');
      mapDispatchToProps(dispatch).toggleChartType('my-chart');
      expect(dispatch.mock.calls).toEqual([
        [
          {
            chartType: 'my-chart',
            requery: 'REQUERY_NEVER',
            type: 'CHART_TYPE_CHANGED',
          },
        ],
      ]);
      expect(gaSpy).toHaveBeenCalledWith('Button', 'Trends:my-chart');
    });
  });

  describe('mapStateToProps', () => {
    it('maps state and props', () => {
      const state = {
        trends: {
          chartType: 'foo',
        },
      };
      const actual = mapStateToProps(state);
      expect(actual).toEqual({ chartType: 'foo' });
    });
  });
});
