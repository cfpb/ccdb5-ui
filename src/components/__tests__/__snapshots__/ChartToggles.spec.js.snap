// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: ChartToggles initial state renders without crashing 1`] = `
<section
  className="chart-toggles m-btn-group"
>
  <p>
    Chart type
  </p>
  <button
    className="a-btn toggle line selected"
    onClick={[Function]}
  >
    <svg
      data-name="line chart"
      id="line-chart-icon"
      viewBox="0 0 27 21"
      xmlns="http://www.w3.org/2000/svg"
    >
      <title>
        line chart icon
      </title>
      <g
        className="cls-1"
      >
        <line
          className="cls-2"
          x1="12"
          x2="2"
          y1="6.27"
          y2="19"
        />
        <line
          className="cls-2"
          x1="22.9"
          x2="14.1"
          y1="14"
          y2="6"
        />
        <path
          className="cls-3"
          d="M13,4a1,1,0,1,1-1,1,1,1,0,0,1,1-1m0-1a2,2,0,1,0,2,2,2,2,0,0,0-2-2Z"
        />
        <path
          className="cls-3"
          d="M24,14a1,1,0,1,1-1,1,1,1,0,0,1,1-1m0-1a2,2,0,1,0,2,2,2,2,0,0,0-2-2Z"
        />
      </g>
      <line
        className="cls-2"
        x1="12"
        x2="2"
        y1="16.45"
        y2="11"
      />
      <line
        className="cls-2"
        x1="23"
        x2="14"
        y1="6"
        y2="16"
      />
      <path
        className="cls-3"
        d="M13,16a1,1,0,1,1-1,1,1,1,0,0,1,1-1m0-1a2,2,0,1,0,2,2,2,2,0,0,0-2-2Z"
      />
      <path
        className="cls-3"
        d="M24,4a1,1,0,1,1-1,1,1,1,0,0,1,1-1m0-1a2,2,0,1,0,2,2,2,2,0,0,0-2-2Z"
      />
    </svg>
  </button>
  <button
    className="a-btn toggle area"
    onClick={[Function]}
  >
    <svg
      data-name="area chart"
      id="area-chart-icon"
      viewBox="0 0 27 21"
      xmlns="http://www.w3.org/2000/svg"
    >
      <title>
        area chart icon
      </title>
      <polygon
        className="cls-1"
        points="27 11 23.48 14 15.26 9 7.04 17 0 15 0 21 27 21 27 11"
      />
      <g
        className="cls-2"
      >
        <polygon
          className="cls-1"
          points="6.32 14.79 14.92 6.46 23.36 11.58 27 8.5 27 1 23.48 4 15.26 1 7.04 7 0 9 0 12.94 6.32 14.79"
        />
      </g>
    </svg>
  </button>
</section>
`;
